Sources/PackageExtension/PackageExtension.swift

```diff
@@ -0,0 +1,339 @@
+    
+    /// Creates a command to run during the build. The executable should be a
+    /// path returned by `TargetBuildContext.lookupTool(named:)`, and all the
+    /// paths in both the command line and the input and output lists should be
+    /// based on the paths in the target build context structure.
+    ///
+    /// Note that input and output dependencies are ignored for prebuild and
+    /// postbuild actions, since they always run before and after the build
+    /// respectively.
+    public func createCommand(displayName: String?, executable: Path, arguments: [String], workingDirectory: Path? = nil, environment: [String: String]? = nil, inputPaths: [Path] = [], outputPaths: [Path] = [], derivedSourcePaths: [Path] = []) {
+        output.commands.append(Command(executable: executable, arguments: arguments, workingDirectory: workingDirectory, environment: environment, displayName: displayName, inputPaths: inputPaths, outputPaths: outputPaths, derivedSourcePaths: derivedSourcePaths))
+    }
+    
+    func addGeneratedOutputFile(path: Path) {
+        output.generatedFilePaths.append(path.string)
+    }
+    
+    public func addPrebuildOutputDirectory(path: Path) {
+        output.prebuildOutputDirectories.append(path.string)
+    }
+}
+
+/// A constructed command.
+fileprivate struct Command: Encodable {
+    let executable: Path
+    let arguments: [String]
+    let workingDirectory: Path?
+    let environment: [String: String]?
+    let displayName: String?
+    let inputPaths: [Path]
+    let outputPaths: [Path]
+    let derivedSourcePaths: [Path]
+}
+
+/// Emits errors, warnings, and remarks to be shown as a result of running the
+/// extension. After emitting one or more errors, the extension should return a
+/// non-zero exit code.
+public final class DiagnosticsEmitter {
+    
+    fileprivate init() {}
+    
+    public func fatalError(_ message: String, file: Path? = nil, line: Int? = nil) {
+        emit(error: message, file: file, line: line)
+        exit(1)
+    }
+    
+    public func emit(error message: String, file: Path? = nil, line: Int? = nil) {
+        output.diagnostics.append(Diagnostic(severity: .error, message: message, file: file, line: line))
+    }
+    public func emit(warning message: String, file: Path? = nil, line: Int? = nil) {
+        output.diagnostics.append(Diagnostic(severity: .warning, message: message, file: file, line: line))
+    }
+    public func emit(remark message: String, file: Path? = nil, line: Int? = nil) {
+        output.diagnostics.append(Diagnostic(severity: .remark, message: message, file: file, line: line))
+    }
+}
+
+// FIXME: This is preliminary.
+public struct Diagnostic: Encodable {
+    enum Severity: String, Encodable {
+        case error, warning, remark
+    }
+    let severity: Severity
+    let message: String
+    let file: Path?
+    let line: Int?
+}
+
+// FIXME: This is preliminary.
+public struct Path: ExpressibleByStringLiteral, Encodable, Decodable {
```

Comment:
it is unfortunate we need to define such in SwiftPM. would adding a dependency on https://github.com/apple/swift-system help?

Reply:
Yes, this is really unfortunate.  The issue is that the PackageExtension API has to be self-contained, like PackageDescription.  If SwiftSystem becomes like Foundation or StdLib so that in be just imported (I don't know if that's the plan) then we could rely on it.  But we should at least make sure the API is the same.  Would it be awful to embed the same implementation?